src/snowflake/cli/api/utils/rendering.py

```diff
@@ -88,24 +97,152 @@ def jinja_render_from_file(
         print(rendered_result)
 
 
-class _AttrGetter:
-    def __init__(self, data_dict):
-        self._data_dict = data_dict
+def _add_project_context(
+    external_data: Dict, project_definition: ProjectDefinition
+) -> Dict:
+    """
+    Updates the external data with variables from snowflake.yml definition file.
+    """
+    if _CONTEXT_KEY in external_data:
+        raise ClickException(
+            f"{_CONTEXT_KEY} in user defined data. The `{_CONTEXT_KEY}` variable is reserved for CLI usage."
+        )
+
+    context_data = _resolve_variables_in_project(project_definition)
+
+    context_data.update(external_data)
```

Comment:
Is this correct direction of updating? Should it really be external -> context, not context -> external? IMO method comment suggests the second direction.

Reply:
I think direction doesn't matter because external_data is at a separate level in this case (top level), while the rest is under ctx.env. My understanding, external_data is only there for backward compatibility with snow sql, and will not be used for non sql stuff. @sfc-gh-turbaszek correct me if I'm wrong.

Reply:
Correct, and in current implementation there's no option for external data to override context data (ctx key is reserved), so direction does not matter.